common/lib/xmodule/xmodule/x_module.py

```diff
@@ -1821,11 +1823,61 @@ def __getattr__(self, name):
         DescriptorSystem, instead. This allows XModuleDescriptors that are bound as XModules
         to still function as XModuleDescriptors.
         """
+        # First we try a lookup in the module system...
         try:
             return getattr(self._module_system, name)
         except AttributeError:
+            pass
+
+        # Before we look into the descriptor system, we're going to check to see
+        # if we're asking for handler_url or local_resource_url, which can be
+        # globally monkeypatched (Studio has to do this because runtimes are
+        # created from modulestores). For more background and refactoring plan,
+        # please see:
+        #
+        # https://openedx.atlassian.net/wiki/display/PLAT/Convert+from+Storage-centric+runtimes+to+Application-centric+runtimes
+        #
+        # The collect phase of BlockStructureTransformers (and 
+        # StudentViewTransformer in particular) only look at the descriptors
+        # (i.e. the parts not bound to a student). In this, it's like Studio.
+        # However, it needs to be able to create LMS runtime URLs (e.g.
+        # transcript download handlers).
+        #
+        # Because I don't really understand the full consequences of
+        # monkey-patching this on the LMS side, I'm going to detect when we have
+        # a url-related request and have *not* monkey-patched the global url
+        # handlers. In that case, we'll create our own LmsHandlerUrls object
+        # manually, and use it to do URL generation.
+        #
+        # The import is embedded here because we're in common/lib and adding the
+        # LmsHandlerUrls introduces a Django dependency.
+        #
+        # This is ugly.
+        is_unpatched = (
+            getattr(descriptor_global_handler_url, 'unpatched', False) and
+            getattr(descriptor_global_local_resource_url, 'unpatched', False)
+        )
+
+        # Force the LMS handler only if all of the following are true:
+        # 1. We're asking for a URL handler.
+        # 2. Nothing has monkey-patched the global values (e.g. Studio).
+        # 3. There is a course_id in the descriptor_system (not true for
+        #    MakoDescriptorSystem).
+        force_lms_url_handler = name in ["handler_url", "local_resource_url"] and is_unpatched
```

Comment:
Good point. I'll try that out. Where do you think it makes sense to apply that monkey patching? I think Studio just picked a module that wouldn't be imported during LMS execution. :-P

Maybe I should just make it a config setting and patch on startup in a consistent way for Studio and LMS?